### PR TITLE
Feature/wfa ca ucc support

### PR DESCRIPTION
**Pull request by morantr**
Add UCC Listener mode to beerocks_cli that listen WFA-CA commands from
the network, and send the replies back to the sender.

In order to use the new UCC listener mode, execute the cli with
argument '-u <port>', for example:
./beerocks_cli -u 7001
The UCC listener will listen to incoming TCP connections on this port
and reply to WFA-CA commands.

As part of this PR, fixed small bugs and remove old WFA-CA old test files.

#121

### CODE REVIEW

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (tomereli) [inline controller/src/beerocks/master/wfa_ca.cpp]: You can just implement it and return "prplMesh-v1.0" or something..

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_bml.cpp:829]: The logic to check for pending response is SO complex and made me read this 5 times till I understood what you're doing here. Its a good sign that this should be done differently.
Why not change the callback to pass `pending_response` instead of this hack style parsing of the reply buffer? it makes no sense to parse it here, this function should be a simple pipe.

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (tomereli) [inline controller/src/beerocks/cli/beerocks_cli_main.cpp:243]: This should be done in a separate commit OR at least add this to the commit message.

**Review by tomereli — DISMISSED** [review]
Now this is a real problem - and why we need design reviews before implementations - when we'll get to agent certification, there will be no controller, hence no BML - so the whole CLI solution you did here can't be used. Please correct me if i'm wrong.

Also I don't think we should remove the script since it might still be useful in the future.

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (morantr) [reply controller/src/beerocks/master/wfa_ca.cpp]: Not in the scope it the task, but ok i'll do it.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp:829]: Or maybe it is because you review it late at night 😉 
Even if I'll pass it in the callback, this piece of code will just move to somewhere else.
At first, I passed it on the BML WFA-CA response message but you guys asked me to remove it.
Basically what I did here is to check if the response message stat with "status,RUNNING"
Maby I will add a comment that explains what I did here, what do think?

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_main.cpp:243]: I did mention it in the commit message, look at the three last lines in the commit message.

**Comment by morantr** [comment]
> Now this is a real problem - and why we need design reviews before implementations - when we'll get to agent certification, there will be no controller, hence no BML - so the whole CLI solution you did here can't be used. Please correct me if i'm wrong.
> 
> Also I don't think we should remove the script since it might still be useful in the future.

The BML supports sending messages to the agent. It doesn't have many commands to send (I think the onbarding query is only one). But when we will get to this point we could add '-r' for a role, or maybe do some command processing to figure out if it is controller or agent.

File: controller/src/beerocks/cli/beerocks_cli_bml.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_bml.cpp:829]: Added a comment

**Review by tomereli — COMMENTED** [review]
Can we not remove the old wfa_ca code just yet? it has been used previously and worked, so its good to keep it as reference.
I'm not approving yet so @arnout has a chance to also review.

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (arnout) [inline controller/src/beerocks/master/wfa_ca.cpp:187]: Doesn't this remove the last *space* (because above, we append `", "`)?

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_main.cpp]: This will break the handling of quoted arguments in parsedCommandArgumentsLinux, won't it?

In addition, I don't see why this would be needed... According to the commit message:

> The non-interactive mode was able to receive only command with no
>     arguments (such as 'bml_conn_map'), 

Actually, I use commands with arguments in non-interactive mode all the time, e.g.  `beerocks_cli  -c "dump_node_info 00:11:22:33:00:20"`. It works for me...

> since the arguments string were
>     cutoff by getopt(). To force getopt() to take the argument one must
>     wrap the command and arguments with single or double-quotes.
>     But then it caused the command to not be found since it has quotes
>     note on its beginning.

Err, no it doesn't... The quotes are interpreted by the shell, and the shell passes the unquoted argument (containing a space) in argv.

>     The solution is to trim these quotes on non-interactive mode.

In other words, I don't see how/why this is needed.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: Fixed a spelling mistake here.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp:15]: NLF usually means Newline - Line Feed, but OK.

However, why is the flushing needed? This is used only on std::cout, as far as I can see, and cout should already be flushed automatically on endl...

**Comment by tomereli** [comment]
@arnout can we merge this PR? we're waiting for your review.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp:106]: TBH I think this approach of a custom message handler is a hack. Basically what you're trying to do is to reuse the listen/accept logic from socket_thread, but without the CMDU handling. Which is essentially a SocketServer. So I think it's better to use SocketServer directly instead of adding this custom_message_handler to socket_thread.

Ideally, there shouldn't even be a need to create a thread for this. You should just be able to add this SocketServer to BML's polling loop. Unfortunatley, the BML interface doesn't allow that. So a new thread is indeed required.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: Does UCC ever put quotes around commands? I don't think this is needed or even appropriate.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [inline controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: `? true : false` makes no sense. If you need an explicit boolean conversion, use `bool (ret == BML_RET_OK)` but that's really not needed here, so just `return (ret == BML_RET_OK)`.

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (arnout) [inline controller/src/beerocks/master/wfa_ca.cpp]: This and many of the following lines had a spurious space at the end of the line. Please configure your editor to avoid that.

End-of-line space is bad because it can easily lead to spurious differences in later edits.

I've fixed this up while rebasing.

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (arnout) [inline controller/src/beerocks/master/wfa_ca.cpp]: This should be indented one space more to align the asterisks.

I've fixed this up while rebasing.

**Comment by arnout** [comment]
> Can we not remove the old wfa_ca code just yet? it has been used previously and worked, so its good to keep it as reference.

I disagree. Dead code is not a good reference.

**Review by arnout — CHANGES_REQUESTED** [review]
The use of socket_thread for something which is not a CMDU communication is a major deficiency IMO. I therefore won't do what I proposed before (just apply and fix up in follow-up), but keep it as changes requested.

However, I did rebase with some minor fixups (spelling mistakes, whitespace errors).

I also properly split the first two commits in 5 commits, like they should have been. It took me 10 minutes, but I think it's worth the time investment. Note that I believe the commit "fix wfa-ca command on non interactive mode" is incorrect and should not be included, but I leave it up to @morantr to confirm that.

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (morantr) [reply controller/src/beerocks/master/wfa_ca.cpp:187]: yes, need to pop 2 time. will fix

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_main.cpp]: It won't break quoted arguments because it only trims on the sides.
> This will break the handling of quoted arguments in parsedCommandArgumentsLinux, won't it?
> 
> In addition, I don't see why this would be needed... According to the commit message:
> 
> > The non-interactive mode was able to receive only command with no
> > arguments (such as 'bml_conn_map'),
>

Exactly, because you used quotes. If you  weren't, then optarg was equal to 'dump_node_info', without the rest of the command.
> Actually, I use commands with arguments in non-interactive mode all the time, e.g. `beerocks_cli -c "dump_node_info 00:11:22:33:00:20"`. It works for me...
> 
> > since the arguments string were
> > cutoff by getopt(). To force getopt() to take the argument one must
> > wrap the command and arguments with single or double-quotes.
> > But then it caused the command to not be found since it has quotes
> > note on its beginning.
> 
AFAIR, when I tested it there were quote characters on optarg.
> Err, no it doesn't... The quotes are interpreted by the shell, and the shell passes the unquoted argument (containing a space) in argv.
> 
> > ```
> > The solution is to trim these quotes on non-interactive mode.
> > ```
> 
> In other words, I don't see how/why this is needed.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp:15]: For debug reasons. Since it is a multithreaded file, I wanted to make sure that the prints will be flushed.

I looked here and it doesn't say explicitly that std::endl flushes the output.
https://en.cppreference.com/w/cpp/io/manip/endl

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp:106]: Agreed on the daily meeting to split it socket thread to cmdu_socket_thread which will be derived from socket thread.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: I think, because I didn't trim it at first, and Tomer asked me to add it when it worked with UCC.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: I have no idea how I wrote such an ugly as that. I will fix.

File: controller/src/beerocks/master/wfa_ca.cpp
Comment (morantr) [reply controller/src/beerocks/master/wfa_ca.cpp]: I don't have such a setting in my editor

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp:15]: Turns out I was wrong - it is std::cerr that is flushed automatically, and not just on endl but on any operation: https://en.cppreference.com/w/cpp/io/manip/unitbuf

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (tomereli) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: I don't recall that I asked that...

Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: So what did you asked me to change? I don't remember either.
Maybe \r\n? because it also trim these characters as well.
So maybe we could change it to:
`beerocks::string_utils::trim(command_string);`

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp:106]: Created #178 to do this in the future.

File: controller/src/beerocks/cli/beerocks_cli_main.cpp
Comment (arnout) [reply controller/src/beerocks/cli/beerocks_cli_main.cpp]: I will remove this.

File: controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp
Comment (arnout) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: I will remove this.

Comment (morantr) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: but not completely, right? only the second argument because trimming the \r\n is required.

Comment (arnout) [reply controller/src/beerocks/cli/beerocks_cli_ucc_listener.cpp]: Thanks for reminding me!

**Review by arnout — APPROVED** [review]

**Pull request merged**
